### PR TITLE
Fix invalid error when the error type reference in the error constructor is loaded via the BIR

### LINKED ISSUE #38346: [Bug]: "cannot create new error value" error when the error type reference in the error constructor is loaded via the BIR
### Description

$title.

Workaround is to use the type in the contextually-expected type.

### Steps to Reproduce

maryam/myproj

```ballerina
public type E1 distinct error;

public type E2 distinct (E1 & error<record {|anydata body;|}>);
```

Usage

```ballerina
import maryam/myproj;

public function main() {
    myproj:E2 e = error("", body = ""); // OK
    error e2 = error myproj:E2("", body = ""); // compilation error: cannot create a new error value from 'myproj:E2'
}
```

### Affected Version(s)

2201.2.1
2201.3.0-SNAPSHOT

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #38346

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -3440,7 +3440,11 @@ public void visit(BLangErrorConstructorExpr errorConstructorExpr, AnalyzerData d
 
         List<BType> errorDetailTypes = new ArrayList<>(expandedCandidates.size());
         for (BType expandedCandidate : expandedCandidates) {
-            BType detailType = ((BErrorType) Types.getReferredType(expandedCandidate)).detailType;
+            BType detailType = Types.getReferredType(expandedCandidate);
+            if (detailType.tag == TypeTags.INTERSECTION) {
+                detailType = ((BIntersectionType) detailType).effectiveType;
+            }
+            detailType = ((BErrorType) detailType).detailType;
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3447]: Can a non error type reach as `detailType` here?

Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3447]: No. But intersection type with error can reach

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -3647,6 +3651,9 @@ private List<BType> getTypeCandidatesForErrorConstructor(BLangErrorConstructorEx
         } else {
             // if `errorTypeRef.type == semanticError` then an error is already logged.
             BType errorType = Types.getReferredType(errorTypeRef.getBType());
+            if (errorType.tag == TypeTags.INTERSECTION) {
+                errorType = ((BIntersectionType) errorType).effectiveType;
+            }
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3656]: Instead, you can use the `getEffectiveType` function of the `Types` class.

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/error_constructor_test.bal [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chiranSachintha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/error_constructor_test.bal:1]:
```suggestion
// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -3464,7 +3464,11 @@ public void visit(BLangErrorConstructorExpr errorConstructorExpr, AnalyzerData d
 
         List<BType> errorDetailTypes = new ArrayList<>(expandedCandidates.size());
         for (BType expandedCandidate : expandedCandidates) {
-            BType detailType = ((BErrorType) Types.getReferredType(expandedCandidate)).detailType;
+            BType detailType = Types.getReferredType(expandedCandidate);
+            if (detailType.tag == TypeTags.INTERSECTION) {
+                detailType = ((BIntersectionType) detailType).effectiveType;
+            }
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3470]: Here also, you can use the `getEffectiveType` function of the `Types` class.

**Review by pcnfernando — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 5]
```diff
@@ -3647,6 +3651,9 @@ private List<BType> getTypeCandidatesForErrorConstructor(BLangErrorConstructorEx
         } else {
             // if `errorTypeRef.type == semanticError` then an error is already logged.
             BType errorType = Types.getReferredType(errorTypeRef.getBType());
+            if (errorType.tag == TypeTags.INTERSECTION) {
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3654]: Why do we need this check?

Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Updated, Please check.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -3464,7 +3464,9 @@ public void visit(BLangErrorConstructorExpr errorConstructorExpr, AnalyzerData d
 
         List<BType> errorDetailTypes = new ArrayList<>(expandedCandidates.size());
         for (BType expandedCandidate : expandedCandidates) {
-            BType detailType = ((BErrorType) Types.getReferredType(expandedCandidate)).detailType;
+            BType detailType = Types.getReferredType(expandedCandidate);
+            detailType = Types.getEffectiveType(detailType);
+            detailType = ((BErrorType) detailType).detailType;
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3469]:
```java
BType detailType = ((BErrorType) Types.getEffectiveType(Types.getReferredType(expandedCandidate))).detailType;
```

Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3469]: This approach decrease the readability of the code ryt?
wdyt?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3469]: No, IMO here the refactored version eliminates some unnecessary variables assignments which can make the code easier to follow. (we can see more clearly the sequence of operations being performed on the `expandedCandidate` variable.)

Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3469]: ok, Will update

Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Updated

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 7]
```diff
@@ -3647,6 +3649,7 @@ private List<BType> getTypeCandidatesForErrorConstructor(BLangErrorConstructorEx
         } else {
             // if `errorTypeRef.type == semanticError` then an error is already logged.
             BType errorType = Types.getReferredType(errorTypeRef.getBType());
+            errorType = Types.getEffectiveType(errorType);
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Updated

**Review by chiranSachintha — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Review by pcnfernando — APPROVED** [review]

**Comment by pcnfernando** [comment]
Send this fix to 3.x and 4.x branches

**Comment by SasinduDilshara** [comment]
> > Fix invalid error when the error type reference in the error constructor is loaded via the BIR.

#39952 and #39953

**Pull request merged**